### PR TITLE
[WAF] precise protection rule change logic parameter to string

### PR DESCRIPTION
**Pull request by cgroschupp**
## Summary of the Pull Request
the logic parameter is a string and no longer an int.

BREAKING CHANGE
It is possible to make it compatible, but I need feedback if that is necessary.

**Remove the `replace` in go.mod before merging.**
## PR Checklist

* [x] Refers to: #1186
* [x] Documentation updated.
* [x] Schema updated.

### CODE REVIEW

File: opentelekomcloud/services/waf/resource_opentelekomcloud_waf_preciseprotection_rule_v1.go [hunk 1]
```diff
@@ -74,9 +75,12 @@ func ResourceWafPreciseProtectionRuleV1() *schema.Resource {
 							ForceNew: true,
 						},
 						"logic": {
-							Type:     schema.TypeInt,
+							Type:     schema.TypeString,
 							Required: true,
 							ForceNew: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								return common.ValidateStringList(v, k, []string{"contain", "not_contain", "equal", "not_equal", "prefix", "not_prefix", "suffix", "not_suffix"})
+							},
```
Comment (outcatcher) [inline opentelekomcloud/services/waf/resource_opentelekomcloud_waf_preciseprotection_rule_v1.go]: There is `validation.StringInSlice` for such cases, please use it.

**Review by outcatcher — CHANGES_REQUESTED** [review]
@cgroschupp 

I suppose it would be better to make a breaking change in the next release, we can increase the minor version (to `v1.25.0`) and schedule it to 14.07. But if it wasn't working, it's just a bug fix, not a breaking change.

Also, I would suggest updating `go.mod` after https://github.com/opentelekomcloud/gophertelekomcloud/pull/206 is merged. 

We can also make a same-day-release of 1.25.0 with backward compatibility breaking in case this topic is really urgent.

**Comment by cgroschupp** [comment]
> @cgroschupp
> 
> I suppose it would be better to make a breaking change in the next release, we can increase the minor version (to `v1.25.0`) and schedule it to 14.07. But if it wasn't working, it's just a bug fix, not a breaking change.
> 
> Also, I would suggest updating `go.mod` after [opentelekomcloud/gophertelekomcloud#206](https://github.com/opentelekomcloud/gophertelekomcloud/pull/206) is merged.
> 
> We can also make a same-day-release of 1.25.0 with backward compatibility breaking in case this topic is really urgent.

I think it had never worked, but I'm not sure. The logic type was wrong and the category in the acceptance test was wrong.

**Comment by outcatcher** [comment]
> > @cgroschupp
> > I suppose it would be better to make a breaking change in the next release, we can increase the minor version (to `v1.25.0`) and schedule it to 14.07. But if it wasn't working, it's just a bug fix, not a breaking change.
> > Also, I would suggest updating `go.mod` after [opentelekomcloud/gophertelekomcloud#206](https://github.com/opentelekomcloud/gophertelekomcloud/pull/206) is merged.
> > We can also make a same-day-release of 1.25.0 with backward compatibility breaking in case this topic is really urgent.
> 
> I think it had never worked, but I'm not sure. The logic type was wrong and the category in the acceptance test was wrong.

I suppose if there was no issue before, it was not used much. We can consider this just a fix without much risk.

**Comment by outcatcher** [comment]
Let's wait for opentelekomcloud/gophertelekomcloud#206 to be merged to get rid of `replace` in dependencies :)

**Comment by cgroschupp** [comment]
@outcatcher done

**Comment by lego963** [comment]
@cgroschupp can you run acceptance test?

**Review by lego963 — APPROVED** [review]

**Comment by cgroschupp** [comment]
@lego963 
```sh
TF_ACC=1 go test -run TestAccWafPreciseProtectionRuleV1_basic -v -timeout 720m ./opentelekomcloud/acceptance/waf/...
=== RUN   TestAccWafPreciseProtectionRuleV1_basic
--- PASS: TestAccWafPreciseProtectionRuleV1_basic (11.71s)
PASS
ok  	github.com/opentelekomcloud/terraform-provider-opentelekomcloud/opentelekomcloud/acceptance/waf	12.438s
```

**Review by outcatcher — APPROVED** [review]

**Pull request merged**